rustls/examples/internal/test_ca.rs

```diff
@@ -0,0 +1,321 @@
+        alg.output_directory()
+            .join(format!("{}.der", self.label()))
+    }
+
+    fn label(&self) -> &'static str {
+        match self {
+            Self::Client => "client",
+            Self::EndEntity => "end",
+            Self::Intermediate => "inter",
+            Self::TrustAnchor => "ca",
+        }
+    }
+
+    fn params(&self, alg: &'static SigAlgContext) -> CertificateParams {
+        let mut params = CertificateParams::default();
+        params.distinguished_name = self.common_name(alg);
+        params.use_authority_key_identifier_extension = true;
+        let serial = SERIAL_NUMBER.fetch_add(1, Ordering::SeqCst);
+        params.serial_number = Some(SerialNumber::from_slice(&serial.to_be_bytes()[..]));
+
+        match self {
+            Self::TrustAnchor | Self::Intermediate => {
+                params.is_ca = IsCa::Ca(BasicConstraints::Unconstrained);
+                params.key_usages = ISSUER_KEY_USAGES.to_vec();
+                params.extended_key_usages = ISSUER_EXTENDED_KEY_USAGES.to_vec();
+            }
+            Self::EndEntity | Self::Client => {
+                params.is_ca = IsCa::NoCa;
+                params.key_usages = EE_KEY_USAGES.to_vec();
+                params.subject_alt_names = vec![
+                    SanType::DnsName(Ia5String::try_from("testserver.com".to_string()).unwrap()),
+                    SanType::DnsName(
+                        Ia5String::try_from("second.testserver.com".to_string()).unwrap(),
+                    ),
+                    SanType::DnsName(Ia5String::try_from("localhost".to_string()).unwrap()),
+                    SanType::IpAddress(IpAddr::from_str("198.51.100.1").unwrap()),
+                    SanType::IpAddress(IpAddr::from_str("2001:db8::1").unwrap()),
+                ];
+            }
+        }
+
+        // Client certificates additionally get the client auth EKU.
+        if matches!(self, Self::Client) {
+            params.extended_key_usages = vec![ExtendedKeyUsagePurpose::ClientAuth];
+        }
+
+        params
+    }
+
+    fn common_name(&self, alg: &'static SigAlgContext) -> DistinguishedName {
+        let mut distinguished_name = DistinguishedName::new();
+        distinguished_name.push(
+            DnType::CommonName,
+            match self {
+                Self::Client => "ponytown client".to_string(),
+                Self::EndEntity => "testserver.com".to_string(),
+                Self::Intermediate => {
+                    format!("ponytown {} level 2 intermediate", alg.issuer_cn)
+                }
+                Self::TrustAnchor => format!("ponytown {} CA", alg.issuer_cn),
+            },
+        );
+        distinguished_name
+    }
+}
+
+// Note: these are ordered such that the data dependencies for issuance are satisfied.
+const ROLES: [Role; 4] = [
+    Role::TrustAnchor,
+    Role::Intermediate,
+    Role::EndEntity,
+    Role::Client,
+];
```

Comment:
Nit: I guess this should go before the definition of `Role`.

Reply:
Fixed.